jib-core/src/main/java/com/google/cloud/tools/jib/cache/Retry.java

```diff
@@ -0,0 +1,159 @@
+  @Nullable private BooleanSupplier successCondition = null;
+  private int maximumRetries = 5;
+  private Predicate<? super E> retryOnException = ignored -> true; // continue to retry
+  private long sleepMilliseconds = -1; // no sleep
+
+  private Retry(Action<E> action) {
+    this.action = action;
+  }
+
+  /**
+   * Set the success condition: the action will be retried until this condition is true.
+   *
+   * @param successCondition the stop condition
+   * @return the instance for further configuration
+   */
+  public Retry<E> until(BooleanSupplier successCondition) {
+    this.successCondition = successCondition;
+    return this;
+  }
+
+  public Retry<E> maximumRetries(int maximumRetries) {
+    Preconditions.checkArgument(maximumRetries > 0);
+    this.maximumRetries = maximumRetries;
+    return this;
+  }
+
+  /**
+   * Provide a predicate to determine if a thrown exception can be retried.
+   *
+   * @param retryOnException determine if provided exception is retryable.
+   * @return the instance for further configuration
+   */
+  public Retry<E> retryOnException(Predicate<Exception> retryOnException) {
+    this.retryOnException = retryOnException;
+    return this;
+  }
+
+  /**
+   * Set the sleep time between retries.
+   *
+   * @param duration the time to sleep
+   * @param unit the unit of time of duration
+   * @return the instance for further configuration
+   */
+  public Retry<E> sleep(long duration, TimeUnit unit) {
+    Preconditions.checkArgument(duration > 0);
+    this.sleepMilliseconds = unit.convert(duration, TimeUnit.MILLISECONDS);
+    return this;
+  }
+
+  /**
+   * Run the action until it runs successfully, to a {@link #maximumRetries(int) maximum number of
+   * retries} (default: 5). An action is deemed to be successful if there are no exceptions and the
+   * {@link #until(BooleanSupplier) optional success condition} is true. If an exception occurs then
+   * the action will be retried providing {@link #retryOnException(Predicate) the exception is
+   * retryable}.
+   *
+   * @return true if the action was run successfully, or {@code false} if the action was unable to
+   *     complete
+   * @throws E exception thrown during the action
+   */
+  public boolean run() throws E {
+    for (int i = 0; i < maximumRetries; i++) {
+      // sleep between attempts, but not on the first attempt
+      if (i > 0 && sleepMilliseconds >= 0) {
+        try {
+          Thread.sleep(sleepMilliseconds);
+        } catch (InterruptedException ex) {
+          // Restore the interrupted status
+          Thread.currentThread().interrupt();
+          return false;
+        }
+      }
+
+      // Do we need to continue?
+      if (successCondition != null && successCondition.getAsBoolean()) {
+        return true;
+      }
+      try {
+        action.run();
+
+        // no exception: check the stop condition, if provided
+        if (successCondition == null || successCondition.getAsBoolean()) {
```

Comment:
I think the configurable `successCondition` is a bit confusing. For example, this check assumes that if a `successCondition` is not explicitly specified, then the operation is deemed successful. But the check in 137, which will be evaluation right after this, assumes that until the given `successCondition` is satisfied, the operation is not deemed successful.

To put this in another way, say I call

- `until(false)` --> executes the operation 5 times and returns false. This feels right.
- `until(true)` --> never executes the operation. Doesn't feel intuitive due to the prevalent familiarity to the structure `do { } until <condition>`.
- not calling `until()` --> executes the operation once. Perhaps `until(true)` should work in this fashion, i.e., the default stop condition when `until()` is not called should be equal to `until(true)`.

Or should the `Action.run()` interface return Boolean, and have the client of `Retry` take the responsibility to evaluate a terminal condition as part of the action?

Reply:
Maybe `until()` isn't the right name.

The idea is that `until()` provides a test as to whether we're in the desired end-state.  If we're already in that state, then running the action is unnecessary.

We were doing this in `moveIfDoesNotExist()` by [an initial `Files.exists(destination)` check](https://github.com/GoogleContainerTools/jib/blob/85c930fccd4ea95f060a6c308d3c52a56dda0341/jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheStorageWriter.java#L81).

Although I could merge this into the action, it would end up looking like:
```
Retry.action(() -> {
    if(Files.exists(destination)) {
      return true;
    }
    Files.move(source, destination);
    return Files.exists(destination);
  })

Reply:
I can't think of a better wording, and that's not so bad.